### DISCUSSION TITLE
Remove vertical bars in multi-line editing in REPL to make copy-paste easier

### DISCUSSION THREAD

**Discussion by bjornregnell** [scala/scala3#17652] [Feature Requests]
The Scala REPL starts emitting vertical bars when entering multi-line editing (e.g. with incomplete code or Alt+Enter):
```
Welcome to Scala 3.2.1 (17.0.5, Java OpenJDK 64-Bit Server VM).
Type in expressions for evaluation. Or try :help.

scala> (
     | 
     | 
     | 
     | 42)
```
This makes copy-paste from REPL to an editor/IDE/Scastie/REPL very tedious as all the bars needs to be removed. We often paste in forums etc and people cannot easily try things out  by a simple copy paste from the post.

An indented line with no `scala>` prompt ~is~ could be enough as an indication of multi-line editing.

Edit: An idea by @smarter from comments below: use another background color as visual cue for multi-line editing mode,  which does not affect copy-paste.

**Comment by smarter** [comment]
The vertical bar comes from https://github.com/lampepfl/dotty/blob/3d251d6fd1bb1e98d66760ee50b824861a87ca57/compiler/src/dotty/tools/repl/JLineTerminal.scala#L35

**Comment by smarter** [comment]
I don't have a strong opinion either way, but I do worry that removing the visual indicator will confuse users, since they might not realize that the REPL is waiting for them to input more code. We would need to replace it by another visual indicator that doesn't break copy-paste.

**Comment by bjornregnell** [comment]
I guess that just replacing the vertical bar with a space is not enough as the bars are assumed elsewhere, as @som-snytt pointed out, such as error messages, carets etc. Perhaps the .stripMargin needs to be replaced by a smarter .stripSpaceMargin ?

**Comment by smarter** [comment]
another idea would be to use a space with a different background color as a visual indicator since copy-pasting should ignore markup.

**Comment by bjornregnell** [comment]
> since they might not realize that the REPL is waiting

Well that might be so, but I have also seen users confused by the vertical bar. "What is that and why is nothing happening???" can still be a reaction.

> use a space with a different background color

That would be nice as it would not break copy-paste!!

**Comment by som-snytt** [comment]
Color space as secondary prompt is a very clever idea, as we've come to expect from dotty team.

I recall wanting to "compose" in repl like a worksheet and get the result out. Maybe a `:print` command to pretty-print history. It could optionally take a range like history command in scala 2. That would print in copy-paste form without any prompts.

Also if we keep VBAR, it could blink like a cursor on the current line of input, to make it more obvious that it is waiting.

![image](https://user-images.githubusercontent.com/369425/218166600-fcd7ac7a-ec5b-4acb-9abf-9ae4c88623da.png)

**Comment by SethTisue** [comment]
> it could blink

 😱

**Comment by bjornregnell** [comment]
>  it could blink

might be a bit too irritating...

**Comment by som-snytt** [comment]
Another idea: reformat on enter and strip the secondary prompt. Stripping the prompt could be a JLine feature contribution.

**Reply by bjornregnell** [reply]
Reformatting on Enter would give the best of both worlds: a visual cue while editing and a copy-paste-friendly text after editing is done! I like it when you can both eat the cake and have it :)

**Reply by bjornregnell** [reply]
Should we open a new issue with the proposal by @som-snytt ?
